rajawali/src/main/java/org/rajawali3d/Geometry3D.java

```diff
@@ -48,658 +50,607 @@
-		mIndicesInt = null;
-		mIndicesShort = null;
-		setIndices(newIntIndices);
+        //Concatenate the old and new data
+        newVertices = concatAllFloat(mVerticesArray, addVertices);
+        newNormals = concatAllFloat(mNormalsArray, addNormals);
+        newColors = concatAllFloat(mColorsArray, addColors);
+        newTextureCoords = concatAllFloat(mTextureCoordsArray, addTextureCoords);
+        newIntIndices = concatAllInt(mIndicesArray, (int[]) addIndices);
+
+        //Set the new data
+        setVertices(newVertices, true);
+        setNormals(newNormals);
+        setTextureCoords(newTextureCoords);
+        setColors(newColors);
+        setIndices(newIntIndices);
 
         if (createVBOs) {
             //Create the new buffers
             createBuffers();
         }
-	}
-
-	/**
-	 * Sets the data. This methods takes two BufferInfo objects which means it'll use another
-	 * Geometry3D instance's data (vertices and normals). The remaining parameters are arrays
-	 * which will be used to create buffers that are unique to this instance.
-	 * <p>
-	 * This is typically used with VertexAnimationObject3D instances.
-	 *
-	 * @param vertexBufferInfo
-	 * @param normalBufferInfo
-	 * @param textureCoords
-	 * @param colors
-	 * @param indices
+    }
+
+    /**
+     * Sets the data. This methods takes two BufferInfo objects which means it'll use another
+     * Geometry3D instance's data (vertices and normals). The remaining parameters are arrays
+     * which will be used to create buffers that are unique to this instance.
+     * <p>
+     * This is typically used with VertexAnimationObject3D instances.
+     *
+     * @param vertexBufferInfo
+     * @param normalBufferInfo
+     * @param textureCoords
+     * @param colors
+     * @param indices
      * @param createVBOs
-	 * @see VertexAnimationObject3D
-	 */
-	public void setData(BufferInfo vertexBufferInfo, BufferInfo normalBufferInfo,
-			float[] textureCoords, float[] colors, int[] indices, boolean createVBOs) {
-		if(textureCoords == null || textureCoords.length == 0)
-			textureCoords = new float[(mNumVertices / 3) * 2];
-		setTextureCoords(textureCoords);
-		if(colors == null || colors.length == 0)
-			setColors(0xff000000 + (int)(Math.random() * 0xffffff));
-		else
-			setColors(colors);
-		setIndices(indices);
-
-		mVertexBufferInfo = vertexBufferInfo;
-		mNormalBufferInfo = normalBufferInfo;
-
-		mOriginalGeometry = null;
+     *
+     * @see VertexAnimationObject3D
+     */
+    public void setData(BufferInfo vertexBufferInfo, BufferInfo normalBufferInfo,
+                        float[] textureCoords, float[] colors, int[] indices, boolean createVBOs) {
+        if (textureCoords == null || textureCoords.length == 0) {
+            textureCoords = new float[(mNumVertices / 3) * 2];
+        }
+        setTextureCoords(textureCoords);
+        if (colors == null || colors.length == 0) {
+            setColors(0xff000000 + (int) (Math.random() * 0xffffff));
```

Comment:
It might be better to make these constants `OPAQUE` and `WHITE` then you can simply do 

```
OPAQUE | (int) (Math.random() * WHITE);
```

Thinking about that more, I'm not sure what the value is in having a huge range of color possibilities.

```
public static final Random RANDOM = new Random();

...

// Create a 'random' color from a a set of 6 possible color shades of RGB then set it opaque 
OPAQUE | 0xf << RANDOM.nextInt(7);
```